sql/parser/builtins.go

```diff
@@ -963,15 +980,20 @@ func datumToString(datum Datum) (string, error) {
 
 var replaceSubRe = regexp.MustCompile(`\\[&1-9]`)
 
-func regexpReplace(s, pattern, to, sqlFlags string) (Datum, error) {
+func regexpReplace(ctx EvalContext, s, pattern, to, sqlFlags string) (Datum, error) {
 	pattern, global, err := regexpEvalFlags(pattern, sqlFlags)
 	if err != nil {
 		return nil, err
 	}
 
-	patternRe, err := regexp.Compile(pattern)
-	if err != nil {
-		return nil, err
+	var patternRe *regexp.Regexp
+	if patternRe = ctx.ReCache.Lookup(pattern); patternRe == nil {
```

Comment:
The locking issue is kind of subtle; I think it's actually better *not* to hold the lock in between the lookup and update. We will on (hopefully rare) occasion compile the same regexp twice, but I think that's better than increasing contention on the cache's lock.

I think it's probably not worth caching regexps that don't compile. These will always fail early rather than being repeated for a large number of rows.